File: py_pub_sub_pkg/py_pub_sub_pkg/subscriber_hello_ros.py
#!/usr/bin/env python3
# This is a shebang line that tells the operating system to use Python 3 to run this script.

import rclpy
# Import the rclpy library, which is the ROS 2 Python client library.

from rclpy.node import Node
# Import the Node class from rclpy, used to create ROS 2 nodes.

from std_msgs.msg import String
# Import the String message type from the std_msgs package, used to publish and subscribe to string messages.

class SubscriberNode(Node):
    # Define a class that inherits from Node, representing a ROS 2 subscriber node.
    def __init__(self):
        # Constructor for the SubscriberNode class.
        super().__init__('subscriber_node')
        # Call the parent class constructor and name this node 'subscriber_node'.
        
        self.subscription = self.create_subscription(
            String,              # Message type for the subscription (std_msgs/String).
            'chatter',           # Topic name to subscribe to ('chatter').
            self.listener_callback, # Callback function to handle received messages.
            10                   # Queue size for messages (10).
        )
        self.subscription
        # Retain the subscription object to prevent it from being garbage collected.

    def listener_callback(self, msg):
        # Callback function that gets executed when a message is received on the 'chatter' topic.
        self.get_logger().info(f"Received: '{msg.data}'")
        # Log the received message to the console.

def main(args=None):
    # Define the main function.
    rclpy.init(args=args)
    # Initialize the rclpy library.
    
    node = SubscriberNode()
    # Create an instance of the SubscriberNode class.
    
    rclpy.spin(node)
    # Keep the node running and listening for messages.
    
    node.destroy_node()
    # Destroy the node when done to clean up resources.
    
    rclpy.shutdown()
    # Shut down the rclpy library to release resources.

if __name__ == '__main__':
    main()

File: cpp_pub_sub_pkg/src/subscriber_hello_ros.cpp
// Subscriber Node
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

class SubscriberNode : public rclcpp::Node {
public:
    SubscriberNode() : Node("subscriber_node") {
        // Create a subscription to the "chatter" topic
        subscription_ = this->create_subscription<std_msgs::msg::String>(
            "chatter", 10, std::bind(&SubscriberNode::listener_callback, this, std::placeholders::_1));
    }

private:
    void listener_callback(const std_msgs::msg::String::SharedPtr msg) {
        // Log the received message
        RCLCPP_INFO(this->get_logger(), "Received: '%s'", msg->data.c_str());
    }

    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
};

int main(int argc, char **argv) {
    // Initialize the rclcpp library
    rclcpp::init(argc, argv);

    // Create a shared pointer for the subscriber node
    auto node = std::make_shared<SubscriberNode>();

    // Spin the node to keep it running and processing callbacks
    rclcpp::spin(node);

    // Shutdown and clean up resources
    rclcpp::shutdown();
    return 0;
}